---
title: Correcting errors in Oxford Nanopore reads with UMIs
author: Florian Bieberich and Aaron Lun
package: sarlacc
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Correcting errors in ONT reads with UMIs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---


```{r, echo=FALSE, results="hide", message=FALSE}
require(knitr)
opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

```{r setup, echo=FALSE, message=FALSE}
library(sarlacc)
```

# Introduction

The `r Biocpkg("sarlacc")` package is designed to perform error correction on long reads such as those generated with the Oxford Nanpore Technology (ONT).
This is done by considering the unique molecular identifier (UMI) associated with each read.
Reads with the same UMI are groups together and used to construct a consensus sequence, thereby correcting for any sequencing errors unique to each read.
In this manner, we can overcome the higher error rate of ONT data while still retaining the long-read capabilities.

# Setting up some sequence data

To demonstrate the various `r Biocpkg("sarlacc")` functions in this vignette, we will mock up some data and save it as a FASTQ file.
Needless to say, this is not necessary for users with sequencing data from real experiments.

```{r}
source(system.file("scripts", "create_mock_seqs.R", package="sarlacc"))
fastq
```

We load in the sequences using `r Biocpkg("Biostrings")` to create a `QualityScaledDNAStringSet`.

```{r}
input <- readQualityScaledDNAStringSet(fastq)
input
```

Note that if quality information is not available, a `DNAStringSet` will suffice as input to `r Biocpkg("sarlacc")` functions.

# Aligning adaptors

## Identifying alignment parameters

Our first task is to align the adaptors to the end of the reads.
For this mock data set, we have the following adaptor sequences:

```{r}
adaptor1
adaptor2
```

We determine suitable alignment parameters using the `tuneAlignment` function.
This will identify the parameter combination that maximizes the separation between the real alignment scores and those for scrambled input sequences^[Note that, for quality-based alignments, `match` and `mismatch` are not used.].

```{r}
tuning <- tuneAlignment(adaptor1, adaptor2, input)
tuning$parameters
```

We confirm that these values are sensible by examining the distribution of scores for the real alignments compared to the scrambled sequences.

```{r, fig.cap="Distribution of adaptor alignment scores for the real reads (blue) and scrambled sequences (red) using the tuned alignment parameters. Scores for both adaptors in the optimal orientation are added together to give the combined score for each read."}
real.dist <- density(tuning$scores$reads)
scram.dist <- density(tuning$scores$scrambled)
plot(real.dist, xlim=range(real.dist$x, scram.dist$x),
    ylim=range(real.dist$y, scram.dist$y), col="blue")
lines(scram.dist, col="red")
```

## Performing the alignment

We now align the adaptor sequences to all of the reads using the `adaptorAlign` function^[We are using fixed parameters chosen from one run of the `tuneAlignment`. We could also supply the tuned parameters directly but these may vary from run to run.].
For speed, the alignment will only consider the 250 bp on either end of the read, under the assumption that the adaptors should occur at the read ends.
This can be further sped up by distributing jobs across multiple cores with the `BPPARAM=` argument.

```{r}
aln.out <- adaptorAlign(adaptor1, adaptor2, input, gapOpening=4, gapExtension=1)
colnames(aln.out)
```

The output is a `DataFrame` that contains alignment information for each adaptor.
For example, the `adaptor1` field contains a nested `DataFrame` with alignment scores and positions for `adaptor1` on the read:

```{r}
aln.out$adaptor1
```

Note that the positions refer to coordinates on the `aln.out$reads`.
These may be different from `input` as the output `reads` will be reverse-complemented (if necessary) to ensure that adaptor 1 and 2 are located on the 5' and 3' ends respectively. 
Whether or not a read sequence was reverse-complemented is indicataed by the `aln.out$reversed` field:

```{r}
summary(aln.out$reversed)
```

# Removing adaptors

We use the `getScoreThreshold` function to choose an appropriate score threshold for the presence of an adaptor.

```{r}
thresh <- getScoreThresholds(aln.out)
thresh$threshold1
thresh$threshold2
```

We verify that the thresholds are suitable for adaptor 1:

```{r, fig.cap="Distribution of adaptor 1 alignment scores for the real reads (blue) and scrambled sequences (red)."}
real.dist <- density(thresh$scores1$reads)
scram.dist <- density(thresh$scores1$scrambled)
plot(real.dist, xlim=range(real.dist$x, scram.dist$x),
    ylim=range(real.dist$y, scram.dist$y), col="blue")
lines(scram.dist, col="red")
abline(v=thresh$threshold1, col="grey", lty=2)
```

... and for adaptor 2:

```{r, fig.cap="Distribution of adaptor 2 alignment scores for the real reads (blue) and scrambled sequences (red)."}
real.dist <- density(thresh$scores2$reads)
scram.dist <- density(thresh$scores2$scrambled)
plot(real.dist, xlim=range(real.dist$x, scram.dist$x),
    ylim=range(real.dist$y, scram.dist$y), col="blue")
lines(scram.dist, col="red")
abline(v=thresh$threshold2, col="grey", lty=2)
```

Based on these thresholds, we use the `chopReads` function to:

- Retain only those reads that have well-aligned adaptors on both sides^[We'll round the thresholds to fixed values, for simplicity and to simplify reproducibility.].
- Remove the ends of the read sequences that align to the adaptors.

```{r}
chopped <- chopReads(aln.out, 12, 13)
chopped$reads
```

One can use the sequences in `reads` directly for downstream analyses if error correction is not necessary. 

# Defining read groups 

## Overview

We want to define groups of reads that are likely to have originated from the same cDNA molecule.
This requires that (i) the reads have similar sequences, and (ii) the reads have similar UMIs.
Thus, we group together reads by applying thresholds to  the percentage of read sequence identity and to the edit distances between UMIs.

Note that there is always a balance between specificity and sensitivity when choosing thresholds:

- Thresholds that are conservative (i.e., require high similarity) will ensure that each read group only has a single molecule of origin.
However, each molecule of origin may give rise to multiple read groups.
- More relaxed thresholds will ensure that each molecule of origin yields no more than one read group.
However, reads in the same group may not come from the same molecule.

For the purpose of error correction, more conservative thresholds are generally preferable.
We do not want to use reads from multiple molecules for correction, especially if molecules differ in their base composition due to, e.g., allelic variation.
(On the other hand, more relaxed thresholds may be desirable if the aim is to eliminate PCR duplicates.)

## Based on the read sequence

To satisfy the first requirement, we perform pairwise alignments between all reads using the `groupReads` function.
This is based on the _minimap2_ software, which must be installed on the system.

```{r}
pre.groups <- groupReads(chopped$reads, truncate=1000)
summary(lengths(pre.groups))
```

For the sake of speed, we only use the first 1 kbp of each read.
This should be sufficient to obtain a rough measure of similarity - grouping on the full read sequence is unnecessary.
We only group reads together if they share at least 70% sequence identity in their first 500 bp.

## Based on the UMI

We use the `umiExtract` function to extract the UMI sequence for each adaptor alignment^[We store the UMI sequences as a column in `chopped` to keep everything in one place.].
By default, it will guess the UMI position from the location of `N`s in the `adaptor1` sequence, but this can also be manually specified if the adaptor sequence is complex.

```{r}
(chopped$umi1 <- umiExtract(chopped$adaptor1))
```

We apply a threshold on the edit distance between UMI sequences to determine whether two reads have the same UMI.
To obtain an appropriate threshold, we use the `expectedDist` function to compute the distribution of edit distances between (nominally identical) adaptor sequences:

```{r, fig.cap="Distribution of edit distances between identical sequences flanking the UMI on adaptor 1."}
edist <- expectedDist(chopped$adaptor1)
hist(edist, xlab="Edit distances", col="grey80")
```

... and we set the threshold to the median edit distance (or 1, if the median is less than zero):

```{r}
(ethresh <- max(1, median(edist)))
```

We group the UMIs for reads _within_ each level of `pre.groups`:

```{r}
groups <- umiGroup(chopped$umi1, threshold1=ethresh, groups=pre.groups)
summary(lengths(groups))
```

If a second UMI is present (e.g., on the other adaptor), these can be handled by repeating the `umiExtract` and `expectedDist` steps above.
The second set of sequences and threshold can then be supplied as additional arguments to `umiGroup`.

# Performing error correction

We perform multiple sequence alignments (MSAs) for each read group using the `r Biocpkg("muscle")` package.
For large data sets, this is often the most time-consuming step and can be parallelized with `BPPARAM`.

```{r}
msa.out <- multiReadAlign(chopped$reads, groups)
lapply(msa.out$alignments[[1]], subseq, 1, 80)
```

We create consensus sequences from these MSAs, representing the error-corrected sequence.
The quality scores are constructed from the qualities of the individual read sequences.
Higher consensus qualities for a position indicate that many reads are in agreement.

```{r}
cons.out <- consensusReadSeq(msa.out)
cons.out
```

This can be saved to a FASTQ file with `writeXStringSet` for further analyses.

# Session information

```{r}
sessionInfo()
```
